### PR TITLE
introduces http2-hpack from Jetty version 9.4.28.v20200408

### PR DESCRIPTION
**Pull request by shamsimam**
Changes from the Jetty version 9.4.28.v20200408 code include:
- introducing a new Utils.java to manage external API changes
- throwing RuntimeException from org.eclipse.jetty.http2.hpack.HpackEncoder#encode(java.nio.ByteBuffer, org.eclipse.jetty.http.MetaData) to preserve backwards compatible signature

**Related** Waiter PR: https://github.com/twosigma/waiter/pull/1089

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -17,7 +17,8 @@
                  [org.eclipse.jetty/jetty-alpn-openjdk8-client ~jetty-version]
                  [org.eclipse.jetty/jetty-alpn-openjdk8-server ~jetty-version]
                  [org.eclipse.jetty.alpn/alpn-api "1.1.3.v20160715"]
-                 [org.eclipse.jetty.http2/http2-common ~jetty-version]
+                 [org.eclipse.jetty.http2/http2-common ~jetty-version
+                  :exclusions [org.eclipse.jetty.http2/http2-hpack]]
```
Comment (shamsimam) [inline project.clj:21]: Exclude `http2-hpack` from the list of dependencies.

File: src/java/org/eclipse/jetty/http2/hpack/Utils.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import org.eclipse.jetty.util.StringUtil;
+
+/**
+ * Helper class to simplify handling the jetty upgrade.
+ */
+public class Utils {
+    /**
+     * Copied from org.eclipse.jetty.http.HttpField#getLowerCaseName
+     */
+    static String getLowerCaseFieldName(final HttpField httpField) {
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/Utils.java:14]: Introduce helper method that copies code from `HttpField` which has a new method to retrieve the name as lower case.

File: src/java/org/eclipse/jetty/http2/hpack/HpackContext.java [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+        if (size > _maxDynamicTableSizeInBytes) {
+            if (LOG.isDebugEnabled())
+                LOG.debug(String.format("HdrTbl[%x] !added size %d>%d", hashCode(), size, _maxDynamicTableSizeInBytes));
+            _dynamicTable.evictAll();
+            return null;
+        }
+        _dynamicTableSizeInBytes += size;
+        _dynamicTable.add(entry);
+        _fieldMap.put(field, entry);
+        _nameMap.put(Utils.getLowerCaseFieldName(field), entry);
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackContext.java:245]: Change to address API change of `field.getLowerCaseName()`

File: src/java/org/eclipse/jetty/http2/hpack/HpackContext.java [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+                Entry entry = _entries[_offset];
+                _entries[_offset] = null;
+                _offset = (_offset + 1) % _entries.length;
+                _size--;
+                if (LOG.isDebugEnabled())
+                    LOG.debug(String.format("HdrTbl[%x] evict %s", HpackContext.this.hashCode(), entry));
+                _dynamicTableSizeInBytes -= entry.getSize();
+                entry._slot = -1;
+                _fieldMap.remove(entry.getHttpField());
+                String lc = Utils.getLowerCaseFieldName(entry.getHttpField());
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackContext.java:351]: Change to address API change of `field.getLowerCaseName()`

File: src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+                for (HttpField field : fields) {
+                    HttpHeader header = field.getHeader();
+                    if (header != null && IGNORED_HEADERS.contains(header))
+                        continue;
+                    if (header == HttpHeader.TE) {
+                        if (field.contains("trailers"))
+                            encode(buffer, TE_TRAILERS);
+                        continue;
+                    }
+                    String name = Utils.getLowerCaseFieldName(field);
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java:215]: Change to address API change of `field.getLowerCaseName()`

File: src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
+    public boolean isValidateEncoding() {
+        return _validateEncoding;
+    }
+
+    public void setValidateEncoding(boolean validateEncoding) {
+        _validateEncoding = validateEncoding;
+    }
+
+    public void encode(ByteBuffer buffer, MetaData metadata) {
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java:150]: Change to not throw `HpackException` as it changes the signature of the method.
The body of the method has been changed to throw a runtime exception `IllegalStateException`.
I verified that exception thrown from these blocks get reported as `INVALID_ARGUMENT` (which grpc status of `3`) errors.

```
ServerReflectionInfo rpc failed. Error code: 3, message: org.eclipse.jetty.http2.hpack.HpackException$SessionException: Could not hpack encode POST{u=http://127.0.0.4:10500/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo,HTTP/2.0,h=11,cl=-1}, debug in
fo: {"created":"@1588585825.367944000","description":"Error received from peer ipv4:127.0.0.1:9091","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"org.eclipse.jetty.http2.hpack.HpackException$SessionException: Could not hpack encode POST{u=h
ttp://127.0.0.4:10500/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo,HTTP/2.0,h=11,cl=-1}","grpc_status":3}    
```

File: src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+            if (_maxHeaderListSize > 0 && _headerListSize > _maxHeaderListSize) {
+                LOG.warn("Header list size too large {} > {} for {}", _headerListSize, _maxHeaderListSize);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("metadata={}", metadata);
+            }
+
+            if (LOG.isDebugEnabled())
+                LOG.debug(String.format("CtxTbl[%x] encoded %d octets", _context.hashCode(), buffer.position() - pos));
+        } catch (HpackException x) {
+            throw new IllegalStateException(x);
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java:232]: Throw a runtime exception to preserve the backward-compatible method signature.

File: src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+        } catch (Throwable x) {
+            HpackException.SessionException failure = new HpackException.SessionException("Could not hpack encode %s", metadata);
+            failure.initCause(x);
+            throw new IllegalStateException(failure);
```
Comment (shamsimam) [inline src/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java:236]: Throw a runtime exception to preserve the backward-compatible method signature.

File: project.clj [hunk 9]
```diff
@@ -50,4 +51,5 @@
           :defaults {:doc/format :markdown}}
   :source-paths ["src/clj"]
   :java-source-paths ["src/java"]
+  :resource-paths ["src/resources"]
```
Comment (shamsimam) [inline project.clj:54]: Include the `src/resources` directory which now includes an implementation of the `HttpFieldPreEncoder` 'service' from the `http2-hpack` code.

**Review by shamsimam — COMMENTED** [review]
Added comments on changes made for including the jetty code.

**Review by sradack — APPROVED** [review]

**Pull request merged**